File: src/jvmTest/java/org/snakeyaml/engine/usecases/inherited/InheritedMarkTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.usecases.inherited;

import it.krzeminski.snakeyaml.engine.kmp.exceptions.Mark;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.stream.Collectors;

import static it.krzeminski.snakeyaml.engine.kmp.usecases.inherited.InheritedImportUtilsKt.getResource;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@org.junit.jupiter.api.Tag("fast")
public class InheritedMarkTest {

  @Test
  @DisplayName("Marks")
  public void testMarks() {
    String content = getResource("test_mark.marks");
    String[] inputs = content.split("---\n");
    for (int i = 1; i < inputs.length; i++) {
      String input = inputs[i];
      int index = 0;
      int line = 0;
      int column = 0;
      while (input.charAt(index) != '*') {
        if (input.charAt(index) != '\n') {
          line += 1;
          column = 0;
        } else {
          column += 1;
        }
        index += 1;
      }
      List<Integer> inputCodepoints = input.codePoints().boxed().collect(Collectors.toList());
      Mark mark = new Mark("testMarks", index, line, column, inputCodepoints, index);
      String snippet = mark.createSnippet(2, 79);
      assertTrue(snippet.contains("\n"), "Must contain at least one '\n'.");
      assertEquals(snippet.indexOf("\n"), snippet.lastIndexOf("\n"), "Must only have only one '\n'.");
      String[] lines = snippet.split("\n");
      String data = lines[0];
      String pointer = lines[1];
      assertTrue(data.length() < 82, "Mark must be restricted: " + data);
      int dataPosition = data.indexOf("*");
      int pointerPosition = pointer.indexOf("^");
      assertEquals(dataPosition, pointerPosition, "Pointer should coincide with '*':\n " + snippet);
    }
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/usecases/inherited/InheritedMarkTest.kt
package it.krzeminski.snakeyaml.engine.kmp.usecases.inherited

import io.kotest.assertions.withClue
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.comparables.shouldBeLessThan
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import it.krzeminski.snakeyaml.engine.kmp.exceptions.Mark
import it.krzeminski.snakeyaml.engine.kmp.internal.utils.toCodePoints

class InheritedMarkTest: FunSpec({
    test("Marks") {
        val content = getResource("test_mark.marks")
        val inputs = content.split("---\n")
            // Necessary because 'split' returns also what's before the first
            // delimiter, as an empty string.
            .dropWhile { it.isEmpty() }
        inputs.forEach { input ->
            var index = 0
            var line = 0
            var column = 0
            while (input[index] != '*') {
                if (input[index] != '\n') {
                    line++
                    column = 0
                } else {
                    column++
                }
                index++
            }
            val inputCodepoints = input.toCodePoints()
            val mark = Mark(
                name = "testMarks",
                index = index,
                line = line,
                column = column,
                codepoints = inputCodepoints,
                pointer = index,
            )
            val snippet = mark.createSnippet(
                indentSize = 2, maxLength = 79,
            )
            withClue("Must contain at least one '\\n'.") {
                snippet shouldContain "\n"
            }
            withClue("Must only have only one '\n'.") {
                snippet.indexOf("\n") shouldBe snippet.lastIndexOf("\n")
            }
            val lines = snippet.split("\n")
            val data = lines[0]
            val pointer = lines[1]
            withClue("Mark must be restricted: $data") {
                data.length shouldBeLessThan 82
            }
            val dataPosition = data.indexOf("*")
            val pointerPosition = pointer.indexOf("^")
            withClue("Pointer should coincide with '*':\n $snippet") {
                dataPosition shouldBe pointerPosition
            }
        }
    }
})
